# 自研前端异常监控系统：js-hunter（客户端篇）

## 前言
在前面的文章[从零开始部署错误报告系统-Sentry](/blog/#/前端工程/Sentry部署)中，我们介绍了如何`Sentry`的部署以及使用，接下来我们试着实现一个属于我们自己的前端异常监控系统。
完整的项目源码地址：[js-hunter](https://github.com/qsz/js-hunter)。我把项目命名为js猎人，希望它能够像猎人一样，敏锐地发现猎物(异常)

## 前端错误以及错误捕获机制

### 有哪些错误
* Javascript 运行时发生的错误
* 资源加载出错
* Promise reject 时未被处理的错误
* xhr 请求出错
* 跨域 JS 资源中事件报错

### 捕获机制
#### window.onerror、addEventListener监听onerror事件
监听运行时且未被处理的错误（JavaScript 运行时错误）
```javascript
window.addEventListener("error", function (event) {
    console.warn("addEventListener - error:", event.message);
});
```
回调中的event对象，即[ErrorEvent](https://developer.mozilla.org/zh-CN/docs/Web/API/ErrorEvent)，包含的主要信息有：
* message 发生错误的描述信息
* filename 发生错误的脚本文件的文件名
* lineno 错误发生时所在的行号
* colno 错误发生时所在的列号

#### 监听 unhandledrejection 事件
当`Promise`被`reject`且没有`reject`处理器的时候，会触发 [unhandledrejection 事件](https://developer.mozilla.org/zh-CN/docs/Web/API/Window/unhandledrejection_event)
```javascript
window.addEventListener("unhandledrejection", (event) => {
    console.warn(`UNHANDLED PROMISE REJECTION: ${event.reason}`);
});
```
举个例子，下面的代码，由于`Promise`缺少`catch`处理，将会抛错，但在`onerror`事件中，却无法监听到 
```javascript
new Promise((resolve) => {
    resolve();
}).then(() => {
    throw 'promise error';
});

window.onerror = (error) => {
    // 无法捕获到错误
    console.log(error)
}

window.addEventListener("unhandledrejection", (event) => {
    console.warn(`捕获到错误: ${event.reason}`);
});
```

#### 接口请求报错
我们需要监听`XMLHttpRequest`的`loadend`事件，该事件会在请求停止之后被触发。在项目中，我们重写`XMLHttpRequest`的`send`方法
```javascript
/**
 * xhr请求出错
 */
function emitXhrError(): void {
  const open = XMLHttpRequest.prototype.open;
  const send = XMLHttpRequest.prototype.send;

  function openReplacement(...args) {
    const method = args[0];
    const url = args[1];
    // 传递请求相关数据
    this._requestInfo = {
      method, // 请求方法
      url, // 请求地址
    };
    return open.apply(this, args);
  }

  function sendReplacement(...args) {
    // 请求停止之后被触发 (例如，在已经触发“error”，“abort”或“load”事件之后)
    // 封装的绑定事件的方法
    bind(this, 'loadend', () => {
      const requestBody = args[0]; // 请求数据
      const { response, status, _requestInfo } = this;

      // 触发捕获
      emitHandler(EventType.XhrError, {
        method: _requestInfo ? _requestInfo.method : '', // 请求方法
        url: _requestInfo ? _requestInfo.url : '', // 请求地址
        requestBody, // 请求体
        response: typeof response === 'object' ? JSON.stringify(response) : response, // 响应体,
        status, // 响应状态
      });

      this._requestInfo = {};
    });
    return send.apply(this, args);
  }

  XMLHttpRequest.prototype.open = openReplacement;
  XMLHttpRequest.prototype.send = sendReplacement;
}
```
在`emitHandler`方法中，触发异常捕获监听
```javascript
export class HttpCapture extends BaseCapture {
  public type = 'HttpCapture';

  protected register(): void {
    registerHandler({
      type: EventType.XhrError,
      callback: (data: { method: string; url: string; requestBody: any; response: string; status: number }) => {
        const { method, url, requestBody, response, status } = data;
        if (
          status === HttpErrorStatusCodes.Unset ||
          (status >= HttpErrorStatusCodes.BadRequest && status < HttpErrorStatusCodes.InternalServerError) ||
          status >= HttpErrorStatusCodes.InternalServerError
        ) {
          const options = this._client.getOptions();
          const { server } = options;
          if (url && url.search(server) > -1) {
            // 数据上报服务出错,则不上报,防止死循环
            logger.warn(`服务地址【${server}】出错, 数据上报失败`);
            return;
          }

          const requestData = {
            method,
            url,
            body: requestBody,
          };
          // 上报数据
          const event = getCommonEventData({
            eventKey: compressString(url, JSON.stringify(requestData)),
            type: EventType.XhrError,
            message: getEventMessage('xhr请求出错', `请求的接口: ${url}`),
            level: EventLevel.Error,
          });
          event.request = requestData;
          event.response = {
            message: response,
            status,
          };

          // 异常信息上报
          this._client.captureEvent(event);
        }
      },
    });
  }
}
```
我们定义一些请求失败的httpCode，当接口返回的`status`符合失败条件时，将会上报错误信息
```javascript
export enum HttpErrorStatusCodes {
    Unset = 0, // 未发送
    BadRequest = 400, // 客户端响应出错
    InternalServerError = 500, // 服务端响应出错
}

```

#### 静态资源请求报错
当静态资源资源( `<img>`、`<script>`等 )加载失败时，元素会触发error事件，并执行该元素上的onerror()处理函数。但需要注意的时，这些error事件不会向上冒泡到window。因此, 在window.onerror中，不能监听到这些error事件。
举个例子
```javascript
<script>
function scriptErrorHandle() {
    console.log("script元素监听到错误");
}

window.onerror = (error) => {
    console.log("无法监听到错误");
};
</script>

<script src="404.js" onerror="scriptErrorHandle()"></script>
```
我们可以改用`addEventListerner`，将`addEventListerner`的第三个参数设置成`true`，我们就可以在捕获阶段监听到这些error。
```javascript
<script>
function scriptErrorHandle() {
    console.log("script元素监听到错误");
}

window.addEventListener(
    "error",
    (event) => {
        console.warn("addEventListener 捕获阶段监听到错误");
    },
    true
);

window.addEventListener(
    "error",
    (event) => {
        console.warn("addEventListener 冒泡阶段无法监听到错误");
    },
    false
);
</script>

<script src="404.js" onerror="scriptErrorHandle()"></script>
```

#### addEventListener整合js运行时的错误和资源请求报错
由于`addEventListener`既能监听onerror事件，也能捕获到资源请求报错，因此，我们将这两类错误在`addEventListener`中同时处理。同时，为了防止错误信息的重复上报，我们需要对错误信息中的`sourceTarget`进行判断，当类型是`HTMLScriptElement`、`HTMLLinkElement`、`HTMLImageElement`就上报资源请求错误，反之上报js运行错误
```javascript
class GlobalExceptionCapture extends BaseCapture {
  public type = 'GlobalExceptionCapture';

  protected register(): void {
    registerHandler({
      type: EventType.Error,
      callback: (data: {
        sourceTarget?: any;
        message: string;
        filename: string;
        lineno: number;
        colno: number;
        error: any;
        originType: string;
      }) => {
        const { sourceTarget, message, filename, lineno, colno, error, originType } = data;

        // 是否是静态资源请求
        const isSourceTarget =
          sourceTarget instanceof HTMLScriptElement ||
          sourceTarget instanceof HTMLLinkElement ||
          sourceTarget instanceof HTMLImageElement;

        // 上报数据
        let event: IEvent = {};

        if (isSourceTarget) {
          // 整理资源请求错误
          const sourceUrl = sourceTarget.src || sourceTarget.href;
          event = getCommonEventData({
            eventKey: compressString('资源地址', sourceUrl),
            type: EventType.LoadError,
            message: getEventMessage('资源请求失败', `资源地址: ${sourceUrl}`),
            level: EventLevel.Warn,
          });
        } else {
          const options = this._client.getOptions();
          if (!urlFilter(options.allowUrls, filename)) {
            // 域名过滤
            return;
          }

          const errorName = errorNameMap[(error && error.name) || ''] || 'Error';
          // TODO 是否需要用 stack 替换 message + filename + lineno + colno
          //  优点: 包含详细的堆栈信息
          //  缺点: 过长, 是否有必要展示这么多信息?
          // const stack = processStackMsg(error);
          const stack = null;
          const eventMsg = stack || `文件${filename} ${lineno}行 ${colno}列 ${message}`;
          event = getCommonEventData({
            eventKey: compressString(message, `${String(filename)}${String(lineno)}${String(colno)}`),
            type: EventType.JavascriptError,
            message: getEventMessage(errorName, eventMsg),
            level: EventLevel.Error,
          });
        }

        event.originType = originType;
        // 异常信息上报
        this._client.captureEvent(event);
      },
    });
  }
}

```

#### 跨域 JS 资源中事件报错
当js资源和当前的域名不一致时，js资源中的代码报错时，onerror中的报错信息只会展示`Script error`。
我们可以利用`try/catch`来捕获错误信息。
```javascript
/**
 * 跨域静态资源中事件报错
 * 重写事件监听
 * 侵入性较强
 */
function emitScriptEventError(): void {
  const originAddEventListener = EventTarget.prototype.addEventListener;
  EventTarget.prototype.addEventListener = function (type, listener: any, options) {
    const wrappedListener = function (...args) {
      try {
        return listener.apply(this, args);
      } catch (error) {
        emitHandler(EventType.ScriptEventError, {
          error,
          target: EventTarget, // 事件对象
          type, // 事件名称
          handlerName: listener && listener.name ? listener.name : '<anonymous>', // 方法名
        });
      }
    };
    return originAddEventListener.call(this, type, wrappedListener, options);
  };
}

```
在`catch`中拿到具体的错误信息后，交由异常处理方法进行上报
```javascript
export class ScriptCapture extends BaseCapture {
  public type = 'ScriptCapture';

  protected _client: IClient<IBrowserOptions>;

  public register(): void {
    const options = this._client.getOptions();
    if (!options.scriptCapture) {
      return;
    }
    registerHandler({
      type: EventType.ScriptEventError,
      callback: (data: { error: any; target: string; type: string; handlerName: string }) => {
        const { error, target, type, handlerName } = data;
        // 上报数据
        const errorName = errorNameMap[(error && error.name) || ''] || 'Error';
        const stack = processStackMsg(error);

        const eventMsg = stack
          ? `跨域js中事件执行报错 ${stack}`
          : `跨域js中事件执行报错 事件对象:${target} 事件名称:${type} 回调函数名称:${handlerName}`;

        const event: IEvent = getCommonEventData({
          eventKey: compressString('scriptEvent', `${eventMsg}`),
          type: EventType.ScriptEventError,
          message: getEventMessage(errorName, eventMsg),
          level: EventLevel.Error,
        });

        // 异常信息上报
        this._client.captureEvent(event);
      },
    });
  }
}

```

### 手动上报错误信息
除了自动捕获错误并且上报之外，用户还可以手动在任何需要的时候上报错误信息。
以`React`为例
```js
import React from 'react';
import * as jsHunter from 'js-hunter';

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    // 更新 state 使下一次渲染能够显示降级后的 UI
    return { hasError: true };
  }

  componentDidCatch(error, info) {
    // 手动上报信息
    jsHunter.sendEvent({
      message: '上报自定义的数据',
      customInfo: '自定义的其他信息',
      timestamp: Date.now(),
      level: 'info', // error | warn | info
    });
  }

  render() {
    if (this.state.hasError) {
      // 自定义降级后的UI
      return <div>加载失败</div>;
    }

    return React.Children.toArray(this.props.children);
  }
}
```

## 上报错误数据

### 定义错误信息的数据结构
```js
export interface IEvent {
  eventKey?: string; // 用于过滤相同的数据 建议上传
  type?: EventType; // 上报类型 建议上传
  message?: string; // 异常信息文字描述 建议上传
  timestamp?: number; // 异常上报时间戳 建议上传
  level?: EventLevel; // 上报级别 建议上传
  pageUrl?: string; // 页面地址
  userAgent?: string; // 浏览器信息
  request?: {
    // 请求数据
    method?: string;
    url?: string;
    body?: any;
  };
  response?: {
    // 响应数据
    message?: string;
    status?: number;
  };
  originType?: string; // 原生事件类型类型 Event.type
  start_timestamp?: number; // 异常开始时间戳
  customInfo?: string; // 用户自定义的信息

  // TODO List
  breadcrumbs?: IBreadcrumb[]; // 用户行为
  SDKversion?: string; // sdk版本
  env?: 'production' | 'development'; // 执行环境
  system?: string; // 系统信息
  user?: string; // 用户信息
}

// 上报类型
export enum EventType {
  Error = 'Error', // 运行时错误 JavascriptError + LoadError
  JavascriptError = 'JavascriptError', // JavaScript运行时错误
  LoadError = 'LoadError', // 资源请求出错
  PromiseError = 'PromiseError', // Promise reject时未被处理的错误(unhandledrejection)
  XhrError = 'XhrError', // xhr请求异常
  UserError = 'UserError', // 用户自定义上传的错误类型
  ScriptEventError = 'ScriptEventError', // 跨域静态资源中事件报错
  ScriptError = 'ScriptError', // 跨域静态资源中js执行报错
  FetchError = 'FetchError', // TODO fetch请求异常
  ReactError = 'ReactError', // TODO react错误边界
  IframeError = 'IframeError', //  TODO iframe错误
}

// 上报级别
export enum EventLevel {
  Error = 'error',
  Warn = 'warn',
  Info = 'info',
}
```

### 同域上报
```js
function sendRequest(event: IEvent): void {
    const globalObject = getGlobalObject();
    if (globalObject.XMLHttpRequest) {
      const xhr = new XMLHttpRequest();
      const server = this._options ? this._options.server : '';

      // ----------POST x-www-form-urlencoded----------
      // 简单跨域请求(不会触发CORS预检)
      xhr.open('POST', `${server}/request`);
      xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded;charset=UTF-8');
      let body = '';
      Object.keys(event).forEach(function (key) {
        let value = event[key];
        if (typeof value === 'function') {
          return;
        }
        if (typeof value === 'object') {
          value = JSON.stringify(value);
        }
        body += `${key}=${value}&`;
      });
      xhr.send(body);
    }
}
```

### 跨域上报
跨域情况下，我们通过请求服务端的一张`request.gif`图片，来上报错误信息
```js
/**
 * 向服务端GET一个1 X 1的gif, 用于上报数据
 * new Image() 不会触发error事件
 */
function _sendImage(event: IEvent): void {
    let img = new Image();
    const server = this._options ? this._options.server : '';
    img.src = `${server}/request.gif?event=${encodeURIComponent(JSON.stringify(event))}`;
    img = null;
}
```

## 构建打包
利用`rollup`打包，支持`ES module`和`script`的方式引入
```js
import typescript from 'rollup-plugin-typescript2';
import commonjs from '@rollup/plugin-commonjs';
import resolve from '@rollup/plugin-node-resolve';
import json from '@rollup/plugin-json';
import alias from '@rollup/plugin-alias'; // 路径别名
import sourceMaps from 'rollup-plugin-sourcemaps';
import { terser } from 'rollup-plugin-terser';
const path = require('path');

const pkg = require('./package.json');
const isProduction = process.env.NODE_ENV === 'production';
const config = {
  input: 'src/index.ts',
  output: [
    { file: pkg.main, format: 'umd', name: 'jsHunter', sourcemap: !!isProduction },
    { file: pkg.module, format: 'es', sourcemap: !!isProduction },
  ],
  watch: {
    include: 'src/**',
  },

  plugins: [
    json(),
    typescript({ useTsconfigDeclarationDir: true }),
    commonjs(),
    resolve(),
    alias({
      entries: {
        '@': path.resolve(__dirname, 'src'),
      },
    }),
  ],
};
if (isProduction) {
  config.plugins.push(sourceMaps());
  config.plugins.push(terser());
}

export default config;

```

### 使用方式
#### JS 项目

```html
<script src="js-hunter.umd.js"></script>
<script src="other.js"></script>
<script>
  jsHunter.init({
    server: 'http://__server__', // 上报的服务器地址
  });

  // 包裹跨域js中的方法, 当方法报错时, 会上报详细信息
  const wrapFn = jsHunter.wrapFnError(otherFn);
  wrapFn();
</script>
```

#### React 项目
项目的入口文件

```js
import * as jsHunter from '@supbd/js-hunter';

jsMonitor.init({
  server: 'http://__server__', // 上报的服务器地址
});
```

## 小结
在这里，我们了解了客户端中会遇到的前端项目错误情况，以及对应的捕获方式，并且通过构建成SDK将错误数据上报。
在下一篇文章中，将尝试封装一个服务端错误收集系统，用于收集、统计客户端上报的错误数据，并用可视化的方式将其展示出来

